certbot/tests/util.py

```diff
@@ -340,8 +341,12 @@ def tearDown(self):
 
         def handle_rw_files(_, path, __):
             """Handle read-only files, that will fail to be removed on Windows."""
-            os.chmod(path, stat.S_IWRITE)
-            os.remove(path)
+            filesystem.chmod(path, stat.S_IWRITE)
+            try:
```

Comment:
No it is not sufficient. The files that are failing here are private keys, that need the full logic of `filesystem` to be available to be processed correctly here.

Reply:
Yes I understand your concern. The comprehensive explanation is quite complex, and I will do a paragraph on that about the spec documentation, but here is a summary if it suits you for now.

The failure is due to the fact that both ownership handling and permission handling need be implemented to have a totally functional logic, and this PR has only the later. 

The reason why we need both is because ownership and permissions are not correlated on windows. This has the consequence that private keys with file mode of type 0400 (read only for user) ends having a list ACL that does not allow the current user, under the administrative privileges, to make modifications on the file, here deleting it.

Once ownership will be implemented, the ACL will be coherent towards this user, as an admin, to effectively delete the file.

Reply:
So, in my previous response, I used neutral and factual statements, but the sense of all of this is that stupid Windows is stupid. Or maybe Python on Windows, or both. But I will say Windows, because Python developers have to deal with the same poor API, like all of us.

The situation we have here occur when we execute Certbot under administrative shell (as it is required for a lot of operations). Let's say we create a file through Python: the root cause is here, instead of having the owner being the actual user (like me, `Adrien Ferrand`), it is the `Administrators` group (that I belong to). First stupidy so.

Let's say now we call our `filesystem.chmod`, that creates the proper ACL list for user, and everyone. The method will get the current user, and construct the ACL list. In fact in this specific case, the byte for user is ignored. Indeed, as the owner is `Administrators`, it is superseeded by the all privileges always given to this group.

So far, so good.

But, for private keys, we use the octal `0o400`. So read for user, nothing for everyone. Here happens the second stupidity. Despite me being part of the `Administrators` group, who has all privileges, Windows fails to say to me that I have in fact all permissions. In the UI for instance, when you check the DACL, Windows says that it cannot show the permissions, you need to explicit invoke elevated privileges to see it.

And here happens the third stupidity: `os.remove` uses the same way to resolve privileges. It does not see the permissions effectively given to user through its membership to `Administrators` group, and so raise a `Permission denied`.

Of course, I could find a way to mitigate that in the tearDown of impacted test classes. First thing I tried was to invoke `filesystem.chmod(path, 0o777)`, to give back permissions to everyone and allow Windows to see the permissions I have on the file. Surprisingly, it does not work, and I still get a permission denied. More precisely, it works afterward, when the Python process finishes, and I am able to remove the file from the UI without any privilege permissions.

I could try to understand why, or find another way to do that, like reimplementing `os.remove` using some magic function in `pywin32` to make low API calls directly with elevated privileges.

But I do not want to do that. Because the real problem here, is that the owner of the file is not the one that is supposed to be. It should be the current user, not the `Administrators` group. Indeed, it raises two absurd situations:
* first, as a user from the `Administrators` group, I cannot manipulate the files for which I have effectively full access to, and provokes the situation we have here with `os.remove`
* second, if I live the `Administrators` group, I lose every access on files on which `0o400` has been applied: despite read permissions have been required for owner, he has none of them ...

So the proper way to solve that, is to ensure that upon file creation, the proper owner is used. This is done in a future PR, that implements the ownership manipulation functions in `filesystem`, and insert in particular a correction of this ownership in the `filesystem.open` and `filesystem.mkdir` functions.

As I said some times, the windows file permissions project is a complete set of actions with a lot of coupling necessary to work correctly (like here, file modes, ownership and creation file). I tried to split it to ease the integration into small PRs, but this is at the cost of breaking temporarily some coupling. Here it is the case.

I hope I answered exhaustively to the situation, and convinced you to just leave this temporary mitigation done in a test class as it is, until the full logic is merged.

Reply:
Thanks for writing all that!

I did some testing myself around this issue and continued to see behavior that surprised me. In particular, I made https://github.com/certbot/certbot/commit/64a1f74df541582b282585c317b152c897031843 based on this branch and [tested it on AppVeyor](https://ci.appveyor.com/project/certbot/certbot/builds/25047231/job/7llg9t88s2n8avip) and the tests passed! Is there an entire other (but very limited) permissions system on Windows we're not properly handling? The behavior I saw here and the second parameter to [CreateFile](https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-createfilea) seems to suggest to me there is.

I'm sorry if you find this frustrating, but I think this PR is the most critical piece of a very important component to Certbot and if we're seeing behavior we don't understand, I'd like us to investigate it to make sure we get this right. Hopefully this comment helps give you an area to look into. I can potentially help with this myself if you want depending on what else I've got going on.

Reply:
Let's clear that thing. In this first comment, I talk about specifications. My second comment will be about the issue that arises here because of this spec.

Indeed, there is a read-only attribute. It is documented here: https://docs.microsoft.com/en-us/windows/desktop/fileio/file-attribute-constants

It is an attribute, not a security flag, and is not part of the security model documented by Microsoft: https://docs.microsoft.com/en-us/windows/desktop/secauthz/access-control

This attribute, as the others (like hidden), can be get or set regarding the values of `FILE_READ_ATTRIBUTES` and `FILE_WRITE_ATTRIBUTES` applied for the current user in respect of the DACL defined for the target file (https://docs.microsoft.com/en-us/windows/desktop/fileio/file-security-and-access-rights)

When set, this attribute, from Python, would raise an `Permission Denied` error if you try to write on the file. You obtain the same error if the DACL of the file does not resolve the `FILE_WRITE_DATA` permission for the current user.

So. I want to step away from this attribute, because:
* it is not part of the security model defined by Microsoft
* it is superseeded on its effects by the definition of the DACL
* how it is handled is done by a different API (`GetFileAttributes`/`SetFileAttributes`), and it does not constitute a security flag, it is controlled by a security flag
* its behavior is inconsistent: in particular, it does not apply to directories, and several reports enlight its confusing behavior (https://support.microsoft.com/en-us/help/326549/you-cannot-view-or-change-the-read-only-or-the-system-attributes-of-fo, https://answers.microsoft.com/en-us/windows/forum/windows_10-files/give-permissions-to-files-and-folders-in-windows/78ee562c-a21f-4a32-8691-73aac1415373).
* its behavior is unsafe from Python, see just below:

Now, about Python. The documentation states that modes have a limited support on Windows: https://docs.python.org/2/library/os.html#os.chmod. In particular, it is stated that only the "read-only flag" is handled by `stat.S_IWRITE` and `stat.S_IREAD` (normally describing r/w permissions on POSIX). It is exactly that: passing these bits to `os.chmod` will enable/disable the flag, not modifying the DACL behind, and so still letting the users with valid rights about attributes (like `Authenticated Users` most of the time) modify the flag.

The only sane way to control the permissions is through the DACL.

Finally, what you are talking about the second parameter in `CreateFile` is not related. It is about the mode of the file descriptor itself (read or read/write), related to `os.O_RDONLY`, `os.O_WDONLY` and `os.O_RDWR` flags passed to `os.open`.

Reply:
So now, @bmw, about the behavior you found on AppVeyor. Congrats, you found a bug in the bug.

If on my personal machine, the owner associated to a file created with `os.open` is `Administrators, it is not the case on AppVeyor and another Windows machine I have (both connected to an Active Directory). For them, the user associated to the file is the real user, not `Administrators`. So, the first bug I exposed here https://github.com/certbot/certbot/pull/6964#discussion_r290483602 does not apply.

Here comes the read-only file attribute. Indeed in the current PR, private keys are typically created using a call like `os.open(path, os.O_CREAT | os.O_EXCL | os.O_RDWR, 0o400)`. Please note the mode `0o400`, that contains `stat.S_IREAD`. So `os.open`, when creating the private key _is_ setting the read-only flag to true.

This leads to the same error than a unsufficient access in the DACL: `Permission denied`.

In your test https://github.com/certbot/certbot/commit/64a1f74df541582b282585c317b152c897031843, you are solving this secondary bug by invoking `os.chmod(path, 0o777)`, whose mode `0o777` contains `stat.S_IWRITE`, and so remove the read-only attribute on this file. Then `os.remove` can work, as the DACL itself contains appropriate ACE to allow the user to delete the file.

Again, this is not something I can solve in this PR. Even if we invoke `os.chmod(path, 0o777)` in the tearDown, we will hit the first bug (improper owner, so improper DACL) in machines where the `Administrators` is the owner of new files.

Both primary bug and secondary bug are solved in the full logic. For the first one (owner problem), I explained that the owner will be consistently on all machines always the real user. 

For the second one, read-only attribute will never be added, because `os.chmod` and `os.open` are forbidden, and replaced by `filesystem.chmod` and `filesystem.open` that never call `os.chmod`.

Reply:
Thanks for the additional explanation. That was very helpful!

While I'm largely satisfied here, one thing still irks me a bit. I cannot reproduce the behavior you described where you are unable to delete the file because it is owned by `Administrators` despite running as a user with `Administrator` privileges invoked. I have manually created files, changed their owner to `Administrator`, and removed all other ACEs, but cannot reproduce the problem. If you're able to give me any instructions on how to do this, I'd certainly appreciate. It'd be useful for me just to verify and learn more about the behavior here.

Finally, while I'll be the first to say it's a bit unnecessary, I'd like to ensure we're not failing to delete files other systems. Can we change the body of the except block to something like:
```
if sys.platform != 'win32':
    raise
```

Reply:
I tested on my personal machine, where Administrators group is set as the owner of the file, and the exception also disappear if we pass a mode to `os.chmod` that removes the read only attribute before trying to delete the file.

Finally it was really about this attribute. I have been confused by the multiple factors involved here, and we can remove a stupidity out of three (which was potentially on python side, so it is great^^). 

We will be cautious, when we will be on the step to add ownership handling, to have a unit test ensuring that the owner is always the actual user, not the group, because there will be no exception raised to warn us about this behavior.

Reply:
Added the conditional raise.